source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.cc

```diff
@@ -0,0 +1,333 @@
+    Runtime::Loader& runtime, Random::RandomGenerator& random,
+    const envoy::config::cluster::v3::Cluster::CommonLbConfig& common_config,
+    const ClientSideWeightedRoundRobinLbProto& client_side_weighted_round_robin_config,
+    TimeSource& time_source)
+    : EdfLoadBalancerBase(
+          priority_set, local_priority_set, stats, runtime, random,
+          PROTOBUF_PERCENT_TO_ROUNDED_INTEGER_OR_DEFAULT(common_config, healthy_panic_threshold,
+                                                         100, 50),
+          LoadBalancerConfigHelper::localityLbConfigFromCommonLbConfig(common_config),
+          /*slow_start_config=*/std::nullopt, time_source) {
+  initialize();
+  orca_load_report_handler_ = std::make_shared<OrcaLoadReportHandler>(
+      client_side_weighted_round_robin_config, time_source_);
+}
+
+// {LoadBalancer} Interface implementation.
+void ClientSideWeightedRoundRobinLoadBalancer::WorkerLocalLb::refreshHostSource(
+    const HostsSource& source) {
+  // insert() is used here on purpose so that we don't overwrite the index if
+  // the host source already exists. Note that host sources will never be
+  // removed, but given how uncommon this is it probably doesn't matter.
+  rr_indexes_.insert({source, seed_});
+  // If the list of hosts changes, the order of picks change. Discard the
+  // index.
+  peekahead_index_ = 0;
+}
+
+HostConstSharedPtr
+ClientSideWeightedRoundRobinLoadBalancer::WorkerLocalLb::chooseHost(LoadBalancerContext* context) {
+  HostConstSharedPtr host = EdfLoadBalancerBase::chooseHost(context);
+  if (context != nullptr) {
+    // Configure callbacks to receive ORCA load report.
+    context->setOrcaLoadReportCallbacks(orca_load_report_handler_);
+  }
+  return host;
+}
+
+double ClientSideWeightedRoundRobinLoadBalancer::WorkerLocalLb::hostWeight(const Host& host) const {
+  ENVOY_LOG(trace, "hostWeight {} = {}", getHostAddress(&host), host.weight());
+  return host.weight();
+}
+
+HostConstSharedPtr ClientSideWeightedRoundRobinLoadBalancer::WorkerLocalLb::unweightedHostPeek(
+    const HostVector& hosts_to_use, const HostsSource& source) {
+  auto i = rr_indexes_.find(source);
+  if (i == rr_indexes_.end()) {
+    return nullptr;
+  }
+  return hosts_to_use[(i->second + (peekahead_index_)++) % hosts_to_use.size()];
+}
+
+HostConstSharedPtr ClientSideWeightedRoundRobinLoadBalancer::WorkerLocalLb::unweightedHostPick(
+    const HostVector& hosts_to_use, const HostsSource& source) {
+  if (peekahead_index_ > 0) {
+    --peekahead_index_;
+  }
+  // To avoid storing the RR index in the base class, we end up using a second
+  // map here with host source as the key. This means that each LB decision
+  // will require two map lookups in the unweighted case. We might consider
+  // trying to optimize this in the future.
+  ASSERT(rr_indexes_.find(source) != rr_indexes_.end());
+  return hosts_to_use[rr_indexes_[source]++ % hosts_to_use.size()];
+}
```

Comment:
Most of these code are duplicate with RoundRobinLoadBalancer. Could our new lb inherit from the round robin directly to deduplicate these code?